### ISSUE TITLE
Add a Report feature to Cascade Tools to track any errors/failures

### ISSUE THREAD

**Issue by HundredVisionsGuy** [HundredVisionsGuy/webcode-tk#22]
### Goal
The primary (and initial) goal of Cascade Tools is to assess color contrast through inheritance. Determine if any element fails color contrast and only includes a list of elements that fail the color contrast algorithm based on the [WebAIM Contrast Checker](https://webaim.org/resources/contrastchecker/)

**User Stories**

- As a teacher grading my students, I want to know which elements failed color contrast and why, so that I can explain to my students how to fix it.
- As a student working on a web project, I want to know if I can get a B or A on the project.
- As a student who failed a color contrast test, I want to know what happened and why as well as how to fix the problem.
- As a student inspecting the color contrast failures, I want to know which files have errors, so I can find and fix the error.
- As a student inspecting the color contrast failures, I want to know which element has the error, so I can fix it.

**Acceptance Criteria**

- Report should only show a list of color contrast failures.
- Report should not overwhelm students with every failed element.
  - Focus on the CSS and not every element
  - Focus on the color rulesets
  - For each color ruleset, check to see if it passes normal AAA and large AAA
    - If it passes normal, then nothing to worry about
    - If it failed large, it would fail everything, we should note that selector and that it fails
    - If it fails normal but not large AND it's targetting headings, then fail for the heading size that is not computed to be large
- Report should be concise and easy to read.
- Report should show the file with the error.
- Report should the element that failed.
- Report should include whether it failed normal or bold text.
- Test should be clearly written and easy to understand
- Message should be no more than 50 characters in length

**Suggestion:
1. GET rating (AA or AAA) to determine level
2. Get global colors for each file

**Report descriptions**

- **Test Wording**: test_project_for_color_contrast (NOTE: this is for the projects - not for cascade_tools)
- **Sample Failure Text Format**: _____ [page] _____ [element] tag fails color contrast for _____ [normal/large]
- **Example**: index.html p tag fails color contrast for normal text

**NOTES**
Going forward, if there are any other tests that require an application of inheritance, we will add it to this report. That would be its own feature request though.

**Comment by HundredVisionsGuy** [comment]
The function: get_color_contrast_details() does the work using a nested function. Here's what it does:

- Takes an AppliedCSSTree object and a rating (default is AAA, and the other option is AA).
- Creates an empty list of results
- Recursively checks all elements to see if they have had color or background color directly applied and for any failure, 
- IF there is a failure
   - it adds a message indicating the file, element, selector, and failure
- It also checks for gradients.
   - If there is a gradient, 
      - it gets all colors and sorts them from lowest contrast ratio to highest.
      - it checks the lowest and the highest, and if there is a failure on either
- At the end of the function, if there are no messages in the list of results, it includes a success message

**Comment by HundredVisionsGuy** [comment]
I got it to work, and you can ignore most of my bullet points from above. It was more like a sound board where I was thinking through the problem in writing. This is what I came up with:
